PowerEditor/src/WinControls/DoubleBuffer/DoubleBuffer.cpp

```diff
@@ -0,0 +1,149 @@
+// This file is part of Notepad++ project
+// Copyright (C) 2024 Jiri Hruska <jirka@fud.cz>
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// at your option any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#include "DoubleBuffer.h"
+#include <CommCtrl.h>
+#include <cassert>
+#include <memory>
+
+void DoubleBuffer::subclass(HWND hWnd)
+{
+	std::unique_ptr<DoubleBuffer> self(new DoubleBuffer);
+	if (::SetWindowSubclass(hWnd, s_subclassWndProc, 0, reinterpret_cast<DWORD_PTR>(self.get())))
+	{
+		self.release();
+	}
+}
+
+DoubleBuffer::~DoubleBuffer()
+{
+	if (_hBufferBitmap)
+	{
+		::SelectObject(_hMemoryDC, _hDefaultBitmap);
+		::DeleteObject(_hBufferBitmap);
+	}
+
+	if (_hMemoryDC)
+	{
+		::DeleteDC(_hMemoryDC);
+	}
+}
+
+void DoubleBuffer::prepareBuffer(HDC hOriginalDC, const RECT& clientRc)
+{
+	if (!_hMemoryDC)
+	{
+		_hMemoryDC = ::CreateCompatibleDC(hOriginalDC);
+	}
+
+	if (!_hBufferBitmap || _size.cx != clientRc.right || _size.cy != clientRc.bottom)
+	{
+		_size = { clientRc.right, clientRc.bottom };
+
+		if (_hBufferBitmap)
+		{
+			::SelectObject(_hMemoryDC, _hDefaultBitmap);
+			::DeleteObject(_hBufferBitmap);
+		}
+
+		_hBufferBitmap = ::CreateCompatibleBitmap(hOriginalDC, _size.cx, _size.cy);
+		if (_hBufferBitmap)
+		{
+			_hDefaultBitmap = static_cast<HBITMAP>(::SelectObject(_hMemoryDC, _hBufferBitmap));
+		}
+	}
+}
+
+LRESULT CALLBACK DoubleBuffer::s_subclassWndProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam, UINT_PTR, DWORD_PTR dwRefData)
```

Comment:
By prefix "s_" in s_subclassWndProc, what does it mean?

Reply:
Yes, correct